### DISCUSSION TITLE
Added: C# Language Design Notes for Mar 21, 2017

### DISCUSSION THREAD

**Discussion by MadsTorgersen** [dotnet/csharplang#641] [General]
[C# Language Design Notes for Mar 21, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-03-21.md)

Discussion of default interface member implementations, based on [this guided tour](https://github.com/dotnet/csharplang/issues/288).

1. Concerns raised on GitHub and elsewhere
2. Inheritance?
3. Breaking name lookup on `this`
4. Events
5. Modifiers
6. Methods
7. Properties
8. Overrides
9. Reabstraction
10. Most specific override
11. Static non-virtual members
12. Accessibility levels
13. Existing programs

Please discuss.

**Comment by jnm2** [comment]
@DavidArno Weren't you just asking for these discussion issues for each meeting notes?

**Comment by DavidArno** [comment]
@jnm2,

Yes, I did and I very much appreciate Mads posting them. Clearly my joking about how this subject probably didn't need more debate missed the mark though. Through my attempts at challenging the team, It appears  that I have poisoned my own position to such an extent that even a quip about such things is seen as me being negative. It's probably long past time for me to move on therefore.

**Comment by jnm2** [comment]
😢 Not what I intended. But perhaps there is a chance you could just avoid the perception by being positive elsewhere? Unless of course you don't want to be.

**Comment by MadsTorgersen** [comment]
@DavidArno these discussion issues are indeed a good idea. Thanks for suggesting it!

**Comment by jamesseanwright** [comment]
**Outsider views klaxon**

I love composition, when it is used pragmatically, so I'm all for [traits](https://github.com/dotnet/csharplang/issues/288). That said, why must this capability be implemented as default `interface` methods? I recall when Java introduced this capability; I still think this adds one concern too many to interfaces.

What about a `trait` keyword?

```csharp
trait Positionable
{
    Point getPosition() { // impl  }
}
```

Perhaps this is pedantic, but I feel a specific keyword would address this responsibility more clearly.

In this case, another word to compose types by `trait`s might be needed for further transparency.

```csharp
class Alien composedof Positionable, Renderable
```

**Comment by HaloFour** [comment]
@jamesseanwright 

The primary motivation of the feature being to allow for the evolution of existing interfaces by adding new non-required (default) methods.  As such it doesn't make a lot of sense to require an interface to have to become something else, even if that something else is just an interface by some other name.  The fact that default method implementations allows for a form of traits is really just a bonus, which is why the LDM is exploring those avenues further.

https://github.com/dotnet/csharplang/issues/52#issuecomment-292248055
https://github.com/dotnet/csharplang/issues/52#issuecomment-306266301

**Comment by jamesseanwright** [comment]
That seems logical; I didn't think of it that way as I'm only using C# with .NET Core for new codebases. Thanks for the reply!
